ml-prims/src/utils.h

```diff
@@ -0,0 +1,212 @@
+class Exception : public std::exception {
+public:
+  /** default ctor */
+  Exception() throw(): std::exception(), msg() {}
+
+  /** copy ctor */
+  Exception(const Exception& src) throw(): std::exception(), msg(src.what()) {
+    collectCallStack();
+  }
+
+  /** ctor from an input message */
+  Exception(const std::string& _msg) throw(): std::exception(), msg(_msg) {
+    collectCallStack();
+  }
+
+  /** dtor */
+  virtual ~Exception() throw() {}
+
+  /** get the message associated with this exception */
+  virtual const char* what() const throw() { return msg.c_str(); }
+
+private:
+  /** message associated with this exception */
+  std::string msg;
+
+  /** append call stack info to this exception's message for ease of debug */
+  // Courtesy: https://www.gnu.org/software/libc/manual/html_node/Backtraces.html
+  void collectCallStack() throw() {
+#ifdef __GNUC__
+    const int MaxStackDepth = 64;
+    void* stack[MaxStackDepth];
+    auto depth = backtrace(stack, MaxStackDepth);
+    std::ostringstream oss;
+    oss << std::endl << "Obtained " << depth << " stack frames" << std::endl;
+    char** strings = backtrace_symbols(stack, depth);
+    if (strings == nullptr) {
+      oss << "But no stack trace could be found!" << std::endl;
+      msg += oss.str();
+      return;
+    }
+    ///@todo: support for demangling of C++ symbol names
+    for (int i = 0; i < depth; ++i) {
+      oss << "#" << i << " in " << strings[i] << std::endl;
+    }
+    free(strings);
+    msg += oss.str();
+#endif // __GNUC__
+  }
+};
+
+/** macro to throw a runtime error */
+#define THROW(fmt, ...)                                                        \
+  do {                                                                         \
+    std::string msg;                                                           \
+    char errMsg[2048];                                                         \
+    std::sprintf(errMsg, "Exception occured! file=%s line=%d: ", __FILE__,     \
+                 __LINE__);                                                    \
+    msg += errMsg;                                                             \
+    std::sprintf(errMsg, fmt, ##__VA_ARGS__);                                  \
+    msg += errMsg;                                                             \
+    throw MLCommon::Exception(msg);                                            \
+  } while (0)
+
+/** macro to check for a conditional and assert on failure */
+#define ASSERT(check, fmt, ...)                                                \
+  do {                                                                         \
+    if (!(check))                                                              \
+      THROW(fmt, ##__VA_ARGS__);                                               \
+  } while (0)
+
+/** check for cuda runtime API errors and assert accordingly */
+#define CUDA_CHECK(call)                                                       \
```

Comment:
I think we should add a CUDA_CHECK_NO_THROW variant of this which only logs the error, but does not throw an exception. This is for example needed in destructors as those should not throw exceptions. To my understanding a prerequisite for this is https://github.com/rapidsai/cuml/issues/100. This is already tracked in https://github.com/rapidsai/cuml/issues/229. So my suggestion would be to add a @todo comment referencing feature request #229 to this pull request.

Reply:
good point. Dropped todo's at relevant places.